forms/templates/forms/form_page.html

```diff
@@ -0,0 +1,41 @@
+{% extends "base.html" %}
+
+{% load wagtailcore_tags %}
+
+{% block main %}
+    <h1 class="page-title">{{ page.title }}</h1>
+
+    <section class="formpage-body">
+        {% if page.intro %}
+            <div class="styled-text">
+                {{ page.intro|richtext }}
+            </div>
+        {% endif %}
+
+        {% if page.form_intro %}
+            <h3 class="heading">{{ page.form_intro }}</h3>
+        {% endif %}
+
+    </section>
+
+    <section class="formpage-form">
+        <form class="form" action="{% pageurl page %}" method="post" novalidate>
+            {% csrf_token %}
+            {% for group in page.groups %}
+                <fieldset class="fieldset fieldset--spaced">
+                    <h4 class="heading-regular">{{ group.title }}</h4>
+                    {% if group.description %}
+                        <p class="hint">{{ group.description }}</p>
+                    {% endif %}
+                    {% with template='forms/_'|add:group.template|add:'.html' %}
+                        {% include template %}
+                    {% endwith %}
```

Comment:
Tell me a little more about your idea here for the `template` data on the `FieldGroup` model. I am wondering if having these pre-defined templates makes form customization a little too inflexible. I think that having field groups is great (and think that nested inlines are great for this purpose). But, and please correct me here if I'm misunderstanding how this is supposed to work, is there something about the design or structure of the form overall that prevents form groups from being more generic container with titles and descriptions that are capable of rendering an arbitrary collection of form fields?

Because from reading the code and playing around a little bit locally, it seems like if I create a group with the "Choices" template and add a URL field to it, the field will not be shown. This particular combination of fields (that is, a choices field and a date field) is not in the design, but I was imagining that if someone _did_ want to make a group like that, it would be possible without adding a new template. If there is a technical limitation preventing this, I'm curious to know what it is.

In general, I do like the idea of giving site editors greater ability to customize the appearance of forms, but I'm feeling like, as implemented here, the templates are in fact restricting what's possible in terms of form creation. Let me know if I'm misunderstanding something, though, as I've definitely had that happen before!

Reply:
@chigby 

I implemented this based on the design - I couldn't see any other way to implement the date without a dedicated template (how would the fields know that they need a forward slash between them/need to be on a single line etc.?).

I separated the templates out based on the design, but I think it's only the date that really needs its own template, the others could use a default loop as you've said.

I was working to the figma design - it would be good to have a design that shows all of the possible implementations (such as select etc.), or these could be removed as options from the edit field if they are not supported?

Reply:
Definitely get what you're saying about matching the design. I'm also a little bit stumped as to how to implement the three year/month/day fields without some kind of workaround. Hmm. So, if I understand correctly, the way to set up a date field that matches the design would be to create a group with the correct template and add three number fields to it, _and_ this type of date field can only exist in a group by itself. 

I mean, I guess the design is the design, but it seems like no matter how we go about this, it's somewhat of a regression in terms of effort and flexibility on the part of the site editor.

Reply:
@chigby 

> So, if I understand correctly, the way to set up a date field that matches the design would be to create a group with the correct template and add three number fields to it, and this type of date field can only exist in a group by itself.

That's the way it currently works, yes.

> I mean, I guess the design is the design, but it seems like no matter how we go about this, it's somewhat of a regression in terms of effort and flexibility on the part of the site editor.

I completely agree - the design doesn't really take into account a content-managed contact form, it's very specific in terms of layout. So I think either the design needs to be amended to work with the grain of a CMS, or a more involved solution is required (which is where we are at currently).

I've pushed some updates today which simplify the available group templates to be either `Default`, which will print out every type of field in order, or `Single date`, which outputs fields inline with a `/` divider. I've also switched to the `sr-only` class for labels, and removed the duplicate class.

That's all I can do for now - I'm away now until I get back to the UK in a few weeks. Hopefully this is now generic _and_ specific enough for the project 🙂